step-functions/step-functions-handler/src/main/java/step/functions/handler/ProcessBasedFunctionHandler.java

```diff
@@ -0,0 +1,268 @@
+        }
+        // Inputs have precedence over properties
+        JsonObject inputJson = (JsonObject) input.getPayload();
+        if (inputJson != null) {
+            for (String key : inputJson.keySet()) {
+                try {
+                    argumentMap.put(key, inputJson.getString(key));
+                } catch (Exception e) {
+                    throw new RuntimeException("Invalid key/value found for key '" + key + "', only string values are supported.");
+                }
+            }
+        }
+        Properties keywordProperties = new Properties();
+        keywordProperties.putAll(argumentMap);
+        // Write properties to file
+        File stepKeywordPropertiesFile = new File(process.getExecutionDirectory(), KEYWORD_INPUT_PROPERTIES_FILE);
+        try (FileOutputStream outputStream = new FileOutputStream(stepKeywordPropertiesFile)) {
+            keywordProperties.store(outputStream, "Keyword properties");
+        }
+    }
+
+    protected static void createProcessInputMessageAsJson(ManagedProcess process, Input<?> input) throws IOException {
+        //Create input properties file from keyword properties and keyword arguments
+        JsonObjectBuilder outputMessageBuilder = Json.createObjectBuilder();
+        outputMessageBuilder.add("inputs", (JsonObject)  input.getPayload());
+        JsonObjectBuilder propertiesJsonObject = Json.createObjectBuilder();
+        input.getProperties().forEach(propertiesJsonObject::add);
+        outputMessageBuilder.add("properties", propertiesJsonObject);
+
+        // Write properties to file
+        File stepKeywordPropertiesFile = new File(process.getExecutionDirectory(), KEYWORD_INPUT_JSON_FILE);
+        try (FileOutputStream outputStream = new FileOutputStream(stepKeywordPropertiesFile);
+             JsonWriter jsonWriter = Json.createWriter(outputStream)) {
+            // Write the JsonObject to the file
+            jsonWriter.writeObject(outputMessageBuilder.build());
+        }
+    }
+
+
+    protected OutputBuilder processOutputAsProperties(ManagedProcess process) {
+        OutputBuilder outputBuilder = new OutputBuilder();
+        File stepKeywordPropertiesFile = new File(process.getExecutionDirectory(), KEYWORD_OUTPUT_PROPERTIES_FILE);
+        if (stepKeywordPropertiesFile.exists() && stepKeywordPropertiesFile.isFile()) {
+            try (FileInputStream fis = new FileInputStream(stepKeywordPropertiesFile)) {
+                Properties outputProperties = new Properties();
+                outputProperties.load(fis);
+                outputProperties.keySet().stream().map(Object::toString).forEach(k -> outputBuilder.add(k,outputProperties.getProperty(k)));
+            } catch (IOException e) {
+                logger.error("Error while reading file " + KEYWORD_OUTPUT_PROPERTIES_FILE, e);
+                outputBuilder.appendError("Error while reading the Keyword output property file. See agent logs for more details.");
```

Comment:
We could also attach the exception to the output. Most users do not have access to the agent logs